File: ios/joinery-kit/Sources/JoineryAIChatKit/ChatAPI.swift
import Foundation
import JoineryKit

/// Thin typed face over the `joinery_ai/chat_*` actions (the native chat
/// surface). Every call rides the app's session key through APIClient;
/// conversations are owner-scoped entirely server-side. Turns run detached on
/// the server, so `send`/`confirm` return a poll handle and `poll` delivers the
/// streaming result — mirroring the web reader's send-then-poll transport.
public struct ChatAPI: Sendable {
    let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /// The caller's conversations, pinned-first then newest, optional search.
    public func list(search: String = "") async throws -> [ChatConversation] {
        var body: [(key: String, value: JSONValue)] = []
        if !search.isEmpty {
            body.append((key: "search", value: .string(search)))
        }
        let envelope = try await client.submitAction("joinery_ai/chat_list", body: .object(body))
        return (envelope["data"]?["conversations"]?.arrayValue ?? []).compactMap { ChatConversation(data: $0) }
    }

    /// One conversation and its turns.
    public func thread(conversationID: Int) async throws -> ChatThreadPayload {
        let envelope = try await client.submitAction("joinery_ai/chat_thread", body: .object([
            (key: "conversation_id", value: .number(Double(conversationID))),
        ]))
        guard let payload = ChatThreadPayload(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return payload
    }

    /// Send a message. Omit `conversationID` to start a new conversation;
    /// `enableDataAccess` seeds a new chat so the assistant can read (and, with
    /// confirmation, write) the owner's data — there's no controls UI yet.
    public func send(message: String, conversationID: Int?, enableDataAccess: Bool) async throws -> ChatSendResult {
        var body: [(key: String, value: JSONValue)] = [
            (key: "message", value: .string(message)),
        ]
        if let conversationID {
            body.append((key: "conversation_id", value: .number(Double(conversationID))))
        } else if enableDataAccess {
            body.append((key: "data_access", value: .bool(true)))
        }
        let envelope = try await client.submitAction("joinery_ai/chat_send", body: .object(body))
        guard let result = ChatSendResult(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return result
    }

    /// One poll tick for a running turn.
    public func poll(messageID: Int) async throws -> ChatPollResult {
        let envelope = try await client.submitAction("joinery_ai/chat_poll", body: .object([
            (key: "message_id", value: .number(Double(messageID))),
        ]))
        guard let result = ChatPollResult(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return result
    }

    /// Resolve a proposed action (`confirm` | `cancel`); the turn resumes on the
    /// same message id, polled like a send.
    public func confirm(conversationID: Int, messageID: Int, decision: String) async throws -> ChatSendResult {
        let envelope = try await client.submitAction("joinery_ai/chat_confirm", body: .object([
            (key: "conversation_id", value: .number(Double(conversationID))),
            (key: "message_id", value: .number(Double(messageID))),
            (key: "decision", value: .string(decision)),
        ]))
        guard let result = ChatSendResult(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return result
    }

    /// Delete a turn; returns the ids the server actually removed (a user turn
    /// takes its paired reply with it).
    @discardableResult
    public func deleteTurn(messageID: Int) async throws -> [Int] {
        let envelope = try await client.submitAction("joinery_ai/chat_turn_action", body: .object([
            (key: "message_id", value: .number(Double(messageID))),
            (key: "action", value: .string("delete")),
        ]))
        return (envelope["data"]?["deleted_ids"]?.arrayValue ?? []).compactMap { $0.intValue }
    }

    public func pin(conversationID: Int, pinned: Bool) async throws {
        try await threadAction("pin", conversationID: conversationID, value: pinned ? "1" : "0")
    }

    public func rename(conversationID: Int, title: String) async throws {
        try await threadAction("rename", conversationID: conversationID, value: title)
    }

    public func deleteConversation(conversationID: Int) async throws {
        try await threadAction("delete", conversationID: conversationID)
    }

    private func threadAction(_ action: String, conversationID: Int, value: String? = nil) async throws {
        var body: [(key: String, value: JSONValue)] = [
            (key: "conversation_id", value: .number(Double(conversationID))),
            (key: "action", value: .string(action)),
        ]
        if let value {
            body.append((key: "value", value: .string(value)))
        }
        _ = try await client.submitAction("joinery_ai/chat_thread_action", body: .object(body))
    }
}

File: android/joinery-android-ai-chat/src/main/java/com/getjoinery/aichat/ChatApi.kt
package com.getjoinery.aichat

import com.getjoinery.android.ApiClient
import com.getjoinery.android.JoineryApiError
import com.getjoinery.android.JsonValue
import com.getjoinery.android.MultipartFile

/**
 * Thin typed face over the `joinery_ai/chat_*` actions (the native chat
 * surface). Every call rides the app's session key through [ApiClient];
 * conversations are owner-scoped entirely server-side. Turns run detached on
 * the server, so `send`/`confirm` return a poll handle and `poll` delivers
 * the streaming result — mirroring the web reader's send-then-poll transport.
 */
class ChatApi(val client: ApiClient) {

    /** The caller's conversations, pinned-first then newest, optional search. */
    suspend fun list(search: String = ""): List<ChatConversation> {
        val body = ArrayList<Pair<String, JsonValue>>()
        if (search.isNotEmpty()) body.add("search" to JsonValue.Str(search))
        val envelope = client.submitAction("joinery_ai/chat_list", JsonValue.Obj(body))
        return (envelope["data"]?.get("conversations")?.arrayValue ?: emptyList())
            .mapNotNull { ChatConversation.from(it) }
    }

    /** One conversation and its turns. */
    suspend fun thread(conversationId: Int): ChatThreadPayload {
        val envelope = client.submitAction(
            "joinery_ai/chat_thread",
            JsonValue.obj("conversation_id" to JsonValue.Num(conversationId.toDouble())),
        )
        return ChatThreadPayload.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Send a message. Omit [conversationId] to start a new conversation, in
     *  which case [seed] carries the control fields the new chat is created
     *  with. When [attachments] is non-empty the call goes out as multipart so
     *  the files reach the server's `$_FILES['attachments']`; otherwise it's a
     *  plain JSON action. */
    suspend fun send(
        message: String,
        conversationId: Int?,
        seed: Map<String, String> = emptyMap(),
        attachments: List<ChatOutgoingAttachment> = emptyList(),
    ): ChatSendResult {
        val envelope: JsonValue
        if (attachments.isEmpty()) {
            val body = ArrayList<Pair<String, JsonValue>>()
            body.add("message" to JsonValue.Str(message))
            if (conversationId != null) {
                body.add("conversation_id" to JsonValue.Num(conversationId.toDouble()))
            } else {
                seed.entries.sortedBy { it.key }.forEach { (key, value) ->
                    body.add(key to JsonValue.Str(value))
                }
            }
            envelope = client.submitAction("joinery_ai/chat_send", JsonValue.Obj(body))
        } else {
            val fields = ArrayList<Pair<String, String>>()
            fields.add("message" to message)
            if (conversationId != null) {
                fields.add("conversation_id" to conversationId.toString())
            } else {
                seed.entries.sortedBy { it.key }.forEach { (key, value) ->
                    fields.add(key to value)
                }
            }
            val files = attachments.map {
                MultipartFile("attachments[]", it.filename, it.mimeType, it.data)
            }
            envelope = client.submitMultipart("joinery_ai/chat_send", fields, files)
        }
        return ChatSendResult.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** The model catalog and default control values for the settings sheet. */
    suspend fun controls(): ChatControlsMeta {
        val envelope = client.submitAction("joinery_ai/chat_controls", JsonValue.Obj(emptyList()))
        return ChatControlsMeta.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Set one control on an existing conversation (string-valued; booleans
     *  as "1"/"0"). Validated server-side against ChatControls. */
    suspend fun setControl(conversationId: Int, field: String, value: String) {
        client.submitAction(
            "joinery_ai/chat_set_capabilities",
            JsonValue.obj(
                "conversation_id" to JsonValue.Num(conversationId.toDouble()),
                "field" to JsonValue.Str(field),
                "value" to JsonValue.Str(value),
            ),
        )
    }

    /** One poll tick for a running turn. */
    suspend fun poll(messageId: Int): ChatPollResult {
        val envelope = client.submitAction(
            "joinery_ai/chat_poll",
            JsonValue.obj("message_id" to JsonValue.Num(messageId.toDouble())),
        )
        return ChatPollResult.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Resolve a proposed action (`confirm` | `cancel`); the turn resumes on
     *  the same message id, polled like a send. */
    suspend fun confirm(conversationId: Int, messageId: Int, decision: String): ChatSendResult {
        val envelope = client.submitAction(
            "joinery_ai/chat_confirm",
            JsonValue.obj(
                "conversation_id" to JsonValue.Num(conversationId.toDouble()),
                "message_id" to JsonValue.Num(messageId.toDouble()),
                "decision" to JsonValue.Str(decision),
            ),
        )
        return ChatSendResult.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Delete a turn; returns the ids the server actually removed (a user
     *  turn takes its paired reply with it). */
    suspend fun deleteTurn(messageId: Int): List<Int> {
        val envelope = client.submitAction(
            "joinery_ai/chat_turn_action",
            JsonValue.obj(
                "message_id" to JsonValue.Num(messageId.toDouble()),
                "action" to JsonValue.Str("delete"),
            ),
        )
        return (envelope["data"]?.get("deleted_ids")?.arrayValue ?: emptyList())
            .mapNotNull { it.intValue }
    }

    suspend fun pin(conversationId: Int, pinned: Boolean) =
        threadAction("pin", conversationId, if (pinned) "1" else "0")

    suspend fun rename(conversationId: Int, title: String) =
        threadAction("rename", conversationId, title)

    suspend fun deleteConversation(conversationId: Int) =
        threadAction("delete", conversationId)

    private suspend fun threadAction(action: String, conversationId: Int, value: String? = null) {
        val body = ArrayList<Pair<String, JsonValue>>()
        body.add("conversation_id" to JsonValue.Num(conversationId.toDouble()))
        body.add("action" to JsonValue.Str(action))
        if (value != null) body.add("value" to JsonValue.Str(value))
        client.submitAction("joinery_ai/chat_thread_action", JsonValue.Obj(body))
    }
}
